src/main/java/pl/psnc/dei/ui/components/PageNavigationComponent.java

```diff
@@ -101,6 +106,26 @@ private void updateElements() {
         add(button);
     }
 
+    /**
+     * blocking last pages if too much pages, performance restriction
+     *
+     * @return
+     */
+    private Button createLastPageButton(int currentPage, int totalPages) {
+        final Button lastPageButton = createButton(totalPages);
+        boolean enableLastPage = shouldEnableLastPage(currentPage, totalPages);
+        lastPageButton.setEnabled(enableLastPage);
+        if (!enableLastPage) {
+            lastPageButton.addClassName("button-disabled");
+        }
+        return lastPageButton;
+    }
+
+    private boolean shouldEnableLastPage(int currentPage, int totalPages) {
+        final int pageDiff = totalPages - currentPage;
+        return (itemsPerPage == 10 && pageDiff < 50) || (itemsPerPage == 20 && pageDiff < 40) || (itemsPerPage == 50 && pageDiff < 30) || (itemsPerPage == 100 && pageDiff < 20);
```

Comment:
I was thinking rather of something like this:

private static final Map<Integer, Integer> THRESHOLDS;

static{
   THRESHOLDS = new HashMap<>();
   THRESHOLDS.put(10,50);
   THRESHOLDS.put(20,40);
   THRESHOLDS.put(50,30);
   THRESHOLDS.put(100,20);
}

and then in the method we have:

private boolean shouldEnableLastPage(int diff) {
  return diff < thresholds.get(itemsPerPage);
}

Isn't that simpler? But I will not insist on my solution... :)